### PR TITLE
Add dockerfile to custom dependencies

### PR DESCRIPTION
**Pull request by priyawadhwa**
Allows users to say something like:

```yaml
- image: my image
  custom:
     buildCommand: ./build.sh
     dependencies:
        dockerfile:
           path: path/to/Dockerfile
           buildArgs:
             - file:foo
           target: secondStage
```

### CODE REVIEW

File: pkg/skaffold/build/custom/dependencies_test.go [hunk 1]
```diff
@@ -25,7 +25,40 @@ import (
...
+	//     file
+	//   Dockerfile
+	tmpDir.Write("foo", "")
+	tmpDir.Write("bar", "")
+	tmpDir.Mkdir("baz")
+	tmpDir.Write("baz/file", "")
+	tmpDir.Write("Dockerfile", "FROM scratch \n ARG file \n COPY $file baz/file .")
+
+	customArtifact := &latest.CustomArtifact{
+		Dependencies: &latest.CustomDependencies{
```
Comment (tejal29) [inline pkg/skaffold/build/custom/dependencies_test.go]: Can you add a test where Target is also specified?

File: pkg/skaffold/build/custom/dependencies.go [hunk 2]
```diff
@@ -26,15 +26,24 @@ import (
...
-	}
-	var dependencies []string
-	for file := range files {
-		dependencies = append(dependencies, file)
+func GetDependencies(ctx context.Context, workspace string, a *latest.CustomArtifact, insecureRegistries map[string]bool) ([]string, error) {
+
+	switch {
+	case a.Dependencies.Dockerfile != nil:
+		dockerfile := a.Dependencies.Dockerfile
+		return docker.GetDependencies(ctx, workspace, dockerfile.Path, dockerfile.BuildArgs, insecureRegistries)
```
Comment (tejal29) [inline pkg/skaffold/build/custom/dependencies.go]: How is a.Dependencies.Target used?

File: pkg/skaffold/schema/latest/config.go [hunk 3]
```diff
@@ -566,14 +566,31 @@ type CustomArtifact struct {
...
 
 // CustomDependencies *alpha* is used to specify dependencies for an artifact built by a custom build script.
+// Either `dockerfile` or `paths` should be specified for file watching to work as expected.
 type CustomDependencies struct {
+	// Dockerfile should be set if the artifact is built from a Dockerfile, from which skaffold can determine dependencies.
+	Dockerfile *DockerfileDependency `yaml:"dockerfile,omitempty" yamltags:"oneOf=dependency"`
 	// Paths should be set to the file dependencies for this artifact, so that the skaffold file watcher knows when to rebuild and perform file synchronization.
 	Paths []string `yaml:"paths,omitempty" yamltags:"oneOf=dependency"`
 	// Ignore specifies the paths that should be ignored by skaffold's file watcher. If a file exists in both `paths` and in `ignore`, it will be ignored, and will be excluded from both rebuilds and file synchronization.
 	// Will only work in conjunction with `paths`.
```
Comment (tejal29) [inline pkg/skaffold/schema/latest/config.go]: Maybe not part of this PR, but we should add validation where skaffold errors when `Ignore` is specified when dependency is dockerfile?
How much ever we specify in docs, users usually don't pay close attention.

File: pkg/skaffold/build/custom/dependencies.go (hunk 2 above)
Comment (priyawadhwa) [reply pkg/skaffold/build/custom/dependencies.go]: sorry, looks like it isn't, i've removed it

File: pkg/skaffold/build/custom/dependencies_test.go (hunk 1 above)
Comment (priyawadhwa) [reply pkg/skaffold/build/custom/dependencies_test.go]: Removed target, as mentioned in comment below!

File: pkg/skaffold/schema/latest/config.go (hunk 3 above)
Comment (priyawadhwa) [reply pkg/skaffold/schema/latest/config.go]: Good point, done!

**Comment by priyawadhwa** [comment]
Yup, we've frozen v1beta9 from the last release and this is making changes to v1beta10.
